<launch>

    <arg name="map_size_x" value="20.0" />
    <arg name="map_size_y" value="10.0" />
    <arg name="map_size_z" value="4.0" />
  
    <arg name="odom_topic" value="odometry" />
    <arg name="drone_id" value="0" />

    <arg name="max_vel" value="3.0"/>
    <arg name="max_acc" value="3.0"/>

    <node pkg="plan_manage" type="topo_traj_node" name="topp_traj_test" output="screen">

        <!-- remap -->
        <remap from="odometry" to="/drone_$(arg drone_id)_odometry"/>

        <!-- particle map -->
        <param name="particle_map/pose_type" value="2" type="int"/>
        <param name="particle_map/frame_id" value="world" type="string"/>
        <param name="particle_map/odom_lidar_timeout" value="0.35" type="double"/>
        <param name="particle_map/local_update_range_x" value="5" type="double"/>
        <param name="particle_map/local_update_range_y" value="5" type="double"/>
        <param name="particle_map/local_update_range_z" value="1.5" type="double"/>
        <param name="particle_map/inf_grid" value="1" type="int"/>
        <param name="particle_map/enable_virtual_wall" value="false" type="bool"/>
        <param name="particle_map/virtual_ceil" type="double" value="10"/>
        <param name="particle_map/virtual_ground" type="double" value="0"/>
        <param name="particle_map/half_fov_horizontal" type="double" value="180"/>
        <param name="particle_map/half_fov_vertical" type="double" value="50"/>
        <param name="particle_map/lidar_topic" type="string" value="/dynamic_map_generator/local_cloud"/>
        <param name="particle_map/pose_topic" type="string" value="pose"/>
        <param name="particle_map/odom_topic" type="string" value="/drone_$(arg drone_id)_odometry"/>
        <param name="particle_map/occupancy_thresh" value="0.5" type="double"/>
        <param name="particle_map/risk_thresh" value = "0.1" type="double"/>
        <param name="particle_map/enable_future_prediction" value="false" type="bool"/>
        <param name="particle_map/prediction_z_height" value="0.2" type="double"/>
        <param name="particle_map/prediction_y_offset" value="25" type="double"/>
        <param name="particle_map/dynamic_cluster_max_center_height" value="20.0" type="double"/>
        <param name="particle_map/dynamic_cluster_max_point_num" value="2008" type="int"/>
        <param name="particle_map/position_prediction_stddev" type="double" value="0.05"/>
        <param name="particle_map/velocity_prediction_stddev" type="double" value="0.05"/>
        <param name="particle_map/new_born_particle_weight" type="double" value="0.0001"/>
        <param name="particle_map/new_born_particle_number_each_point" type="int" value="15"/>
        <param name="particle_map/voxel_filter_resolution" type="double" value="0.2"/>
        <param name="particle_map/sigma_ob" type="double" value="0.1"/>

        <!-- topo prm -->
        <param name="topo_prm/sample_inflate_x" value="1.0" type="double"/>
        <param name="topo_prm/sample_inflate_y" value="3.5" type="double"/>
        <param name="topo_prm/sample_inflate_z" value="1.0" type="double"/>
        <param name="topo_prm/clearance" value="0.3" type="double"/>
        <param name="topo_prm/max_sample_time" value="0.005" type="double"/>
        <param name="topo_prm/max_sample_num" value="2000" type="int"/>
        <param name="topo_prm/max_raw_path" value="300" type="int"/>
        <param name="topo_prm/max_raw_path2" value="25" type="int"/>
        <param name="topo_prm/short_cut_num" value="1" type="int"/>
        <param name="topo_prm/reserve_num" value="4" type="int"/>
        <param name="topo_prm/ratio_to_short" value="5.5" type="double"/>
        <param name="topo_prm/parallel_shortcut" value="false" type="bool"/>
        <param name="topo_prm/risk_thresh" value="0.05" type="double"/>
        <param name="topo_prm/init_x" value="-10" type="double"/>
        <param name="topo_prm/init_y" value="0" type="double"/>
        <param name="topo_prm/init_z" value="2.0" type="double"/>

        <param name="manager/max_vel" value="$(arg max_vel)" type="double" />
        <param name="manager/max_acc" value="$(arg max_acc)" type="double" />
        <param name="manager/max_jerk" value="4" type="double" />
        <param name="manager/control_points_distance" value="0.4" type="double" />
        <param name="manager/feasibility_tolerance" value="0.05" type="double" />
        <param name="manager/planning_horizon" value="4.5" type="double" />
        <!-- <param name="manager/use_distinctive_trajs" value="$(arg use_distinctive_trajs)" type="bool" /> -->
        <param name="manager/topo_max_num" value="10" />

        <param name="manager/drone_id" value="$(arg drone_id)" />
        <!-- <param name="manager/risk_weight" value="100.0" /> -->
        
        <!-- trajectory optimization -->
        <param name="optimization/lambda1_smooth" value="1.0" type="double"/>
        <param name="optimization/lambda1_guide" value="1.5" type="double"/>
        <param name="optimization/lambda1_feasibility" value="0.1" type="double"/>
        
        <param name="optimization/lambda2_smooth" value="1.0" type="double"/>
        <param name="optimization/lambda2_feasibility" value="0.5" type="double"/>
        <param name="optimization/lambda2_fitness" value="1.0" type="double"/>


        <param name="optimization/dist0" value="0.5" type="double"/>
        <param name="optimization/max_vel" value="$(arg max_vel)" type="double"/>
        <param name="optimization/max_acc" value="$(arg max_acc)" type="double"/>

        <param name="bspline/limit_vel" value="$(arg max_vel)" type="double"/>
        <param name="bspline/limit_acc" value="$(arg max_acc)" type="double"/>
        <param name="bspline/limit_ratio" value="1.1" type="double"/>


    </node>

    <include file="$(find plan_manage)/launch/include/map.xml">
        <arg name="map_size_x_" value="$(arg map_size_x)" />
        <arg name="map_size_y_" value="$(arg map_size_y)" />
        <arg name="map_size_z_" value="$(arg map_size_z)" />
        <arg name="odom_topic" value="/drone_$(arg drone_id)_$(arg odom_topic)" />
        <arg name="drone_id" value="$(arg drone_id)" />
    </include>

    <include file="$(find plan_manage)/launch/include/simulator.xml">
        <arg name="drone_id" value="$(arg drone_id)" />
        <arg name="map_size_x_" value="$(arg map_size_x)" />
        <arg name="map_size_y_" value="$(arg map_size_y)" />
        <arg name="map_size_z_" value="$(arg map_size_z)" />
        <arg name="init_x_" value="-10" />
        <arg name="init_y_" value="0" />
        <arg name="init_z_" value="2.0" />
        <arg name="odometry_topic" value="$(arg odom_topic)" />
    </include>

    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find plan_manage)/config/topo.rviz" required="true" />

</launch>